x/mongo/driver/ocsp/ocsp.go

```diff
@@ -0,0 +1,276 @@
+			// one to responsd.
+			httpResponse, err := http.DefaultClient.Do(request)
+			if err != nil {
+				urlErr, ok := err.(*url.Error)
+				if !ok {
+					return nil
+				}
+
+				timeout := urlErr.Timeout()
+				cancelled := urlErr.Err == context.Canceled // Timeout() does not return true for context.Cancelled.
+				if userContextUsed && (timeout || cancelled) {
+					// Handle the original context expiring or being cancelled.
+					return err
+				}
+				return nil // Ignore all other errors.
+			}
+			defer func() {
+				_ = httpResponse.Body.Close()
+			}()
+			if httpResponse.StatusCode != 200 {
+				return nil
+			}
+
+			httpBytes, err := ioutil.ReadAll(httpResponse.Body)
+			if err != nil {
+				return nil
+			}
+
+			ocspResponse, err := ocsp.ParseResponseForCert(httpBytes, cfg.serverCert, cfg.issuer)
+			if err != nil || verifyResponse(cfg, ocspResponse) != nil || ocspResponse.Status == ocsp.Unknown {
+				// If there was an error parsing/validating the response or the response was inconclusive, suppress
+				// the error because we want to ignore this responder.
+				return nil
+			}
+
+			// Store the response and return a sentinel error so the error group will exit and any in-flight requests
+			// will be cancelled.
+			ocspResponses <- ocspResponse
+			return errGotOCSPResponse
+		})
+	}
+
+	if err := group.Wait(); err != nil && err != errGotOCSPResponse {
+		return nil, err
+	}
+	if len(ocspResponses) == 0 {
+		// None of the responders gave a conclusive response.
+		return nil, nil
+	}
+	return <-ocspResponses, nil
+}
+
+// verifyResponse checks that the provided OCSP response is valid. An error is returned if the response is invalid or
+// reports that the certificate being checked has been revoked.
+func verifyResponse(cfg config, res *ocsp.Response) error {
+	if err := verifyExtendedKeyUsage(cfg, res); err != nil {
+		return err
+	}
+
+	currTime := time.Now().UTC()
+	if res.ThisUpdate.After(currTime) {
+		return fmt.Errorf("reported thisUpdate time %s is after current time %s", res.ThisUpdate, currTime)
+	}
+	if !res.NextUpdate.IsZero() && res.NextUpdate.Before(currTime) {
+		return fmt.Errorf("reported nextUpdate time %s is before current time %s", res.NextUpdate, currTime)
+	}
+	return nil
+}
+
+func verifyExtendedKeyUsage(cfg config, res *ocsp.Response) error {
+	if res.Certificate == nil {
+		return nil
+	}
+
+	namesMatch := res.RawResponderName != nil && bytes.Equal(res.RawResponderName, cfg.issuer.RawSubject)
+	keyHashesMatch := res.ResponderKeyHash != nil && bytes.Equal(res.ResponderKeyHash, cfg.ocspRequest.IssuerKeyHash)
+	if namesMatch || keyHashesMatch {
+		// The responder certificate is the same as the issuer certificate.
+		return nil
+	}
+
+	// There is a delegate.
+	var foundExtendedUsage bool
+	for _, extKeyUsage := range res.Certificate.ExtKeyUsage {
+		if extKeyUsage == x509.ExtKeyUsageOCSPSigning {
+			foundExtendedUsage = true
```

Comment:
Nit: Can you return nil here, and not worry about managing the state for foundExtendedUsage?

Reply:
Btw, I'd be totally fine if you changed nothing to Evergreen, those were just stream of concious thoughts as I read the file.

Reply:
This was originally part of the `verifyResponse` function rather than a separate one so it was written with a `break`. Changed to return early instead.